masonry/src/passes/mutate.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use tracing::info_span;
+
+use crate::passes::merge_state_up;
+use crate::render_root::RenderRoot;
+use crate::widget::WidgetMut;
+use crate::{MutateCtx, Widget, WidgetId, WidgetState};
+
+pub(crate) fn mutate_widget<R>(
+    root: &mut RenderRoot,
+    id: WidgetId,
+    mutate_fn: impl FnOnce(WidgetMut<'_, Box<dyn Widget>>) -> R,
+) -> R {
+    let (widget_mut, state_mut) = root.widget_arena.get_pair_mut(id);
+
+    let _span = info_span!("mutate_widget", name = widget_mut.item.short_type_name()).entered();
+    // NOTE - parent_widget_state can be None here, because the loop below will merge the
+    // state up to the root.
+    let root_widget = WidgetMut {
+        ctx: MutateCtx {
+            global_state: &mut root.state,
+            parent_widget_state: None,
+            widget_state: state_mut.item,
+            widget_state_children: state_mut.children,
+            widget_children: widget_mut.children,
+        },
+        widget: widget_mut.item,
+    };
+
+    let result = mutate_fn(root_widget);
+
+    // Merge all state changes up to the root.
+    let mut current_id = Some(id);
+    while let Some(id) = current_id {
+        let parent_id = root.widget_arena.parent_of(id);
+        merge_state_up(&mut root.widget_arena, id);
+        current_id = parent_id;
+    }
+
+    result
+}
+
+// TODO - Add link to mutate pass documentation
+/// Apply any deferred mutations (created using [...Ctx::mutate_later](crate::LayoutCtx::mutate_later)).
+pub(crate) fn run_mutate_pass(root: &mut RenderRoot, root_state: &mut WidgetState) {
+    // TODO - Factor out into a "pre-event" function?
+    // root.state.next_focused_widget = root.state.focused_widget;
+
+    let callbacks = std::mem::take(&mut root.state.mutate_callbacks);
+    for callback in callbacks {
+        mutate_widget(root, callback.id, callback.callback);
+    }
+
+    root_state.merge_up(root.widget_arena.get_state_mut(root.root.id()).item);
+    // root.post_event_processing(&mut root_state);
```

Comment:
We should reuse the allocation here (comment on the implicit `drop(callbacks)`)

Of course, we need to be aware of the fact that new callbacks could have been added, but that won't be the general case.

Reply:
Hard disagree. Skipping one allocation per frame is not worth even one line of code. Especially if we need additional logic in case new callbacks have been added (though I understand you think that shouldn't be possible).

Reply:
It's not necessarily just one allocation, if this feature is used widely. But sure, I'm not going to block on this.

Reply:
Yeah likely more allocations, in case there are added a few callbacks (which seems likely). I don't yet see how your case applies, it's sync isn't it? It could also be double-buffered with a second `Vec` and `std::mem::swap` alternatively. But yeah also no strong feelings here (just seems to be a cheap optimization, which can still be reversed later)...

Reply:
Right.

Even then, `mutate_callbacks` is a Vec of boxes, so we're talking about saving `O(log(N))` reallocs per frame in a context where we have `O(N)` allocations minimum. There's no situation in which that meaningfully impacts performance.

And I think `render_root` is *the* file where you can to keep complexity down as much as you can. It's a complexity bottleneck, so to speak, both because it naturally tends to amass more code as features get added, and because anyone who wants to global view of the code needs to read it.